readme.md
GraphQL HTTP Server Middleware
==============================

Create a GraphQL HTTP server with koa v2. it's ported from [express-graphql](https://github.com/graphql/express-graphql).
koa-graphql-next have the same api with express-graphql, and well tested like express-graphql,that's mean,and could be try
to using in product currently.

## Faq

1. only supported koa v2
2. add url to options,and default url is '/graphql'
3. using jest and supertest,instead of mocha in express-graphql.all test from express-graphql (now 64 tests),have been rewrited and passed.
4. ported from express-graphql,for the code,"line by line",and rewrites all the test of express-graphql,so fix many bug founded in tests.
   but,thanks for the author of these project:
   [koa-graphql by chentsulin](https://github.com/chentsulin/koa-graphql) 
   [graffiti](https://github.com/RisingStack/graffiti) it proveder a middleware for koa v1.
   [koa-graphql by Arch-Mage](https://github.com/Arch-Mage/koa-graphql/tree/next)
5. for the detail of usage,could read the test,for example how to upload files with graphql.

## Installation

```sh
npm install --save koa-graphql-next
```

Then use `koa-graphql-next` at any point as a middleware with koa@next:

```js
import graphqlHTTP from 'koa-graphql-next';

const app = new koa();

app.use( graphqlHTTP({
  schema: MyGraphQLSchema,
  graphiql: true
}));

app.listen(3000);
```

## Options

The `graphqlHTTP` function accepts the following options:

  * **`schema`**: A `GraphQLSchema` instance from [`graphql-js`][].
    A `schema` *must* be provided.

  * **`context`**: A value to pass as the `context` to the `graphql()`
    function from [`graphql-js`][].

  * **`rootValue`**: A value to pass as the `rootValue` to the `graphql()`
    function from [`graphql-js`][].

  * **`pretty`**: If `true`, any JSON response will be pretty-printed.

  * **`formatError`**: An optional function which will be used to format any
    errors produced by fulfilling a GraphQL operation. If no function is
    provided, GraphQL's default spec-compliant [`formatError`][] function will
    be used.

  * **`validationRules`**: Optional additional validation rules queries must
    satisfy in addition to those defined by the GraphQL spec.

  * **`graphiql`**: If `true`, may present [GraphiQL][] when loaded directly
    from a browser (a useful tool for debugging and exploration).

  * **`url`**: added in koa-graphql-next,An optional string for the query url.
    default value is '/graphql'.

## Debugging

During development, it's useful to get more information from errors, such as
stack traces. Providing a function to `formatError` enables this:

```js
formatError: error => ({
  message: error.message,
  locations: error.locations,
  stack: error.stack
})
```


## HTTP Usage

Once installed at a path, `koa-graphql-next` will accept requests with
the parameters:

  * **`query`**: A string GraphQL document to be executed.

  * **`variables`**: The runtime values to use for any GraphQL query variables
    as a JSON object.

  * **`operationName`**: If the provided `query` contains multiple named
    operations, this specifies which operation should be executed. If not
    provided, a 400 error will be returned if the `query` contains multiple
    named operations.

  * **`raw`**: If the `graphiql` option is enabled and the `raw` parameter is
    provided raw JSON will always be returned instead of GraphiQL even when
    loaded from a browser.

GraphQL will first look for each parameter in the URL's query-string:

```
/graphql?query=query+getUser($id:ID){user(id:$id){name}}&variables={"id":"4"}
```

If not found in the query-string, it will look in the POST request body.

If a previous middleware has already parsed the POST body, the `ctx.request.body`
value will be used. Use [`koa-multer`][] or a similar middleware to add support
for `multipart/form-data` content, which may be useful for GraphQL mutations
involving uploading files. 

If the POST body has not yet been parsed, koa-graphql-next will interpret it
depending on the provided *Content-Type* header.

  * **`application/json`**: the POST body will be parsed as a JSON
    object of parameters.

  * **`application/x-www-form-urlencoded`**: this POST body will be
    parsed as a url-encoded string of key-value pairs.

  * **`application/graphql`**: The POST body will be parsed as GraphQL
    query string, which provides the `query` parameter.


[`graphql-js`]: https://github.com/graphql/graphql-js
[`formatError`]: https://github.com/graphql/graphql-js/blob/master/src/error/formatError.js
[GraphiQL]: https://github.com/graphql/graphiql



changelog.md

#20170129 v1.0.6

##update graphql to v0.9.1
##using supertest,remove supertest-as-promised
##add doc for update npm.

#20161115 v1.0.5

## change test framwork from ava to jest.
## update graphql to v0.8.1
  fix only an http error message test

#20160827 v1.0.4
## update graphql to v0.7.0,the release note is here
    https://github.com/graphql/graphql-js/releases/tag/v0.7.0
    and express-graphql add an test for new behaver:
    https://github.com/graphql/express-graphql/commit/536f8653e9e3d8d893ae7a8e02ffed4fe3b4d88b

## context defaulting to request
   a test of session changed to this way.

## Force client to use UTF-8 encoding by setting encoding in Content-Type

## all changes from express0.53 has been review and fix on 1.04.
all commit since express-graphql 0.53 here:
1. [not need]Remove http-errors from devDependencies  
2. [not need]use res.send directly to send data not need
3. [not need]Return main promise not need
4. [not need]change Middleware definition to return promise not need
5. [changed]Force client to use UTF-8 encoding by setting encoding in Content-Type
6. [not need]graphql-express -> express-graphql in readme  
7. [not need]Return main promise PR#99
8. [not need]Remove http-errors from devDependencies
9. [changed]Force client to use UTF-8 encoding by setting encoding in Content-Type
10. [not need]Merge branch 'use-end-to-send-data' 
11. [not need]Merge branch 'chentsulin-use-end-to-send-data'
12. [changed]context defaulting to request

#20160808 v1.0.3
  add test for koa-session and koa-generic.session
  update graphql to v0.6.2
  update babel and other depends to newest.

#v1.0.2
  when using a function as argument,only pass ctx to the function.

#v1.0.1
  clean the package.json

#v1.0.0
  ported from express-graphql.
  test rewrite using ava+supertest-as-promised

src/__test__/http.spec.js
/**
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 */

// 80+ char lines are useful in describe/it, so ignore in this file.
/* eslint-disable max-len */

import { stringify } from 'querystring';
import url from 'url';
import zlib from 'zlib';
//import multer from 'multer';
//import bodyParser from 'body-parser';
import test from 'ava';
import request from 'supertest-as-promised';
import koa from 'koa';

import {
    GraphQLSchema,
    GraphQLObjectType,
    GraphQLNonNull,
    GraphQLString,
    GraphQLError,
    BREAK
} from 'graphql';
import graphqlHTTP from '../';

const QueryRootType = new GraphQLObjectType({
    name: 'QueryRoot',
    fields: {
        test: {
            type: GraphQLString,
            args: {
                who: {
                    type: GraphQLString
                }
            },
            resolve: (root, { who }) => 'Hello ' + (who || 'World')
        },
        thrower: {
            type: new GraphQLNonNull(GraphQLString),
            resolve: () => { throw new Error('Throws!'); }
        },
        context: {
            type: GraphQLString,
            resolve: (obj, args, context) => context,
        }
    }
});

const TestSchema = new GraphQLSchema({
    query: QueryRootType,
    mutation: new GraphQLObjectType({
        name: 'MutationRoot',
        fields: {
            writeTest: {
                type: QueryRootType,
                resolve: () => ({})
            }
        }
    })
});

function urlString(urlParams) {
    let string = '/graphql';
    if (urlParams) {
        string += ('?' + stringify(urlParams));
    }
    return string;
}

function catchError(p) {
    return p.then(
        () => { throw new Error('Expected to catch error.'); },
        error => {
            if (!(error instanceof Error)) {
                throw new Error('Expected to catch error.');
            }
            return error;
        }
    );
}

function promiseTo(fn) {
    return new Promise((resolve, reject) => {
        fn((error, result) => error ? reject(error) : resolve(result));
    });
}

test('test harness', async (t) => {

    let caught;
    try {
        await catchError(Promise.resolve());
    } catch (error) {
        caught = error;
    }
    t.is('Expected to catch error.', caught && caught.message, 'expects to catch errors');

    try {
        await catchError(Promise.reject('not a real error'));
    } catch (error) {
        caught = error;
    }
    t.is('Expected to catch error.', caught && caught.message, 'expects to catch actual errors');

    const resolveValue = {};
    const result = await promiseTo(cb => cb(null, resolveValue));
    t.is(resolveValue, result, 'resolves callback promises');

    const rejectError = new Error();
    try {
        await promiseTo(cb => cb(rejectError));
    } catch (error) {
        caught = error;
    }
    t.is(rejectError, caught, 'rejects callback promises with errors');
});

//start to test GET functionality
test('allows GET with query param', async (t) => {
    const app = new koa();
    app.use(graphqlHTTP({
        schema: TestSchema
    }));
    const server = app.listen();
    let serverRes= await request(server)
        .get(urlString({
            query: '{test}'
        }));

    t.is(serverRes.res.text,'{"data":{"test":"Hello World"}}');
});

    //   it('allows GET with variable values', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({
    //       schema: TestSchema
    //     }));

    //     const response = await request(app)
    //       .get(urlString({
    //         query: 'query helloWho($who: String){ test(who: $who) }',
    //         variables: JSON.stringify({ who: 'Dolly' })
    //       }));

    //     expect(response.text).to.equal(
    //       '{"data":{"test":"Hello Dolly"}}'
    //     );
    //   });

    //   it('allows GET with operation name', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP(() => ({
    //       schema: TestSchema
    //     })));

    //     const response = await request(app)
    //       .get(urlString({
    //         query: `
    //           query helloYou { test(who: "You"), ...shared }
    //           query helloWorld { test(who: "World"), ...shared }
    //           query helloDolly { test(who: "Dolly"), ...shared }
    //           fragment shared on QueryRoot {
    //             shared: test(who: "Everyone")
    //           }
    //         `,
    //         operationName: 'helloWorld'
    //       }));

    //     expect(JSON.parse(response.text)).to.deep.equal({
    //       data: {
    //         test: 'Hello World',
    //         shared: 'Hello Everyone',
    //       }
    //     });
    //   });

    //   it('Reports validation errors', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({ schema: TestSchema }));

    //     const error = await catchError(
    //       request(app)
    //         .get(urlString({
    //           query: '{ test, unknownOne, unknownTwo }'
    //         }))
    //     );

    //     expect(error.response.status).to.equal(400);
    //     expect(JSON.parse(error.response.text)).to.deep.equal({
    //       errors: [
    //         {
    //           message: 'Cannot query field "unknownOne" on type "QueryRoot".',
    //           locations: [ { line: 1, column: 9 } ]
    //         },
    //         {
    //           message: 'Cannot query field "unknownTwo" on type "QueryRoot".',
    //           locations: [ { line: 1, column: 21 } ]
    //         }
    //       ]
    //     });
    //   });

    //   it('Errors when missing operation name', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({ schema: TestSchema }));

    //     const error = await catchError(
    //       request(app)
    //         .get(urlString({
    //           query: `
    //             query TestQuery { test }
    //             mutation TestMutation { writeTest { test } }
    //           `
    //         }))
    //     );

    //     expect(error.response.status).to.equal(400);
    //     expect(JSON.parse(error.response.text)).to.deep.equal({
    //       errors: [
    //         { message: 'Must provide operation name if query contains multiple operations.' }
    //       ]
    //     });
    //   });

    //   it('Errors when sending a mutation via GET', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({ schema: TestSchema }));

    //     const error = await catchError(
    //       request(app)
    //         .get(urlString({
    //           query: 'mutation TestMutation { writeTest { test } }'
    //         }))
    //     );

    //     expect(error.response.status).to.equal(405);
    //     expect(JSON.parse(error.response.text)).to.deep.equal({
    //       errors: [
    //         { message: 'Can only perform a mutation operation from a POST request.' }
    //       ]
    //     });
    //   });

    //   it('Errors when selecting a mutation within a GET', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({ schema: TestSchema }));

    //     const error = await catchError(
    //       request(app)
    //         .get(urlString({
    //           operationName: 'TestMutation',
    //           query: `
    //             query TestQuery { test }
    //             mutation TestMutation { writeTest { test } }
    //           `
    //         }))
    //     );

    //     expect(error.response.status).to.equal(405);
    //     expect(JSON.parse(error.response.text)).to.deep.equal({
    //       errors: [
    //         { message: 'Can only perform a mutation operation from a POST request.' }
    //       ]
    //     });
    //   });

    //   it('Allows a mutation to exist within a GET', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({ schema: TestSchema }));

    //     const response = await request(app)
    //       .get(urlString({
    //         operationName: 'TestQuery',
    //         query: `
    //           mutation TestMutation { writeTest { test } }
    //           query TestQuery { test }
    //         `
    //       }));

    //     expect(response.status).to.equal(200);
    //     expect(JSON.parse(response.text)).to.deep.equal({
    //       data: {
    //         test: 'Hello World'
    //       }
    //     });
    //   });

    //   it('Allows passing in a context', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP({
    //       schema: TestSchema,
    //       context: 'testValue'
    //     }));

    //     const response = await request(app)
    //       .get(urlString({
    //         operationName: 'TestQuery',
    //         query: `
    //           query TestQuery { context }
    //         `
    //       }));

    //     expect(response.status).to.equal(200);
    //     expect(JSON.parse(response.text)).to.deep.equal({
    //       data: {
    //         context: 'testValue'
    //       }
    //     });
    //   });

    //   it('Allows returning an options Promise', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP(() => Promise.resolve({
    //       schema: TestSchema,
    //     })));

    //     const response = await request(app)
    //       .get(urlString({
    //         query: '{test}'
    //       }));

    //     expect(response.text).to.equal(
    //       '{"data":{"test":"Hello World"}}'
    //     );
    //   });

    //   it('Catches errors thrown from options function', async () => {
    //     const app = server();

    //     app.use(urlString(), graphqlHTTP(() => {
    //       throw new Error('I did something wrong');
    //     }));

    //     const req = request(app)
    //       .get(urlString({
    //         query: '{test}'
    //       }));

    //     const error = await catchError(req);

    //     expect(error.response.status).to.equal(500);
    //     expect(error.response.text).to.equal(
    //       '{"errors":[{"message":"I did something wrong"}]}'
    //     );
    //   });
    // });


src/__test__/index.spec.js
import graphqlHTTP from '../';
import test from 'ava';
import request from 'supertest-as-promised';
import koa from 'koa';

//options should be provided
test('requires an option factory function', async (t) => {
    t.throws(() => {
        graphqlHTTP();
    }, 'GraphQL middleware requires options.'); //must throw an exception
});

//options as function
test('requires option factory function to return object', async (t) => {
    const app = new koa();
    app.use(graphqlHTTP(() => null));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    t.is(result.res.statusCode,500);
    t.deepEqual(JSON.parse(result.res.text), {
        errors: [
            {
                message:
                'GraphQL middleware option function must return an options object or a promise which will be resolved to an options object.'
            }
        ]
    });
}); //end function

//options as promise
test('requires option factory function to return object or promise of object', async (t) => {
    const app = new koa();
    app.use(graphqlHTTP(() => Promise.resolve(null)));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    t.is(result.res.statusCode,500);
    t.deepEqual(JSON.parse(result.res.text), {
        errors: [
            {
                message:
                'GraphQL middleware option function must return an options object or a promise which will be resolved to an options object.' 
            }
        ]
    });
}); //end function

test('requires option factory function to return object or promise of object with schema', async (t) => {
    const app = new koa();
    app.use(graphqlHTTP(() => Promise.resolve({})));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    t.is(result.res.statusCode,500);
    t.deepEqual(JSON.parse(result.res.text), {
        errors: [
            {
                message:
                 'GraphQL middleware options must contain a schema.'
            }
        ]
    });
}); //end function

src/__test__/options.spec.js
import graphqlHTTP from '../';
import request from 'supertest';
import koa from 'koa';

//options should be provided
it('requires an option factory function', async () => {
    expect(() => {
        graphqlHTTP();
    }).toThrowError('GraphQL middleware requires options.'); //must throw an exception
});

//options as function
it('requires option factory function to return object', async () => {
    const app = new koa();
    app.use(graphqlHTTP(() => null));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    expect(result.res.statusCode).toBe(500);
    expect(JSON.parse(result.res.text)).toEqual({
        errors: [
            {
                message:
                'GraphQL middleware option function must return an options object or a promise which will be resolved to an options object.'
            }
        ]
    });
}); //end function

//options as promise
it('requires option factory function to return object or promise of object', async () => {
    const app = new koa();
    app.use(graphqlHTTP(() => Promise.resolve(null)));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    expect(result.res.statusCode).toBe(500);
    expect(JSON.parse(result.res.text)).toEqual({
        errors: [
            {
                message:
                'GraphQL middleware option function must return an options object or a promise which will be resolved to an options object.'
            }
        ]
    });
}); //end function

it('requires option factory function to return object or promise of object with schema', async () => {
    const app = new koa();
    app.use(graphqlHTTP(() => Promise.resolve({})));
    const server = app.listen();
    let result = await request(server).get('/graphql?query={test}');
    expect(result.res.statusCode).toBe(500);
    expect(JSON.parse(result.res.text)).toEqual({
        errors: [
            {
                message:
                'GraphQL middleware options must contain a schema.'
            }
        ]
    });
}); //end function

src/parseBody.js
/**
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 */

import contentType from 'content-type';
import getBody from 'raw-body';
import httpError from 'http-errors';
import querystring from 'querystring';
import zlib from 'zlib';

export function parseBody(ctx) {
    return new Promise((resolve, reject) => {

        //here,when we use other body parser at first,we should work normally
        //express body parse test ok,but in koa,koa-bodyparse couldn't work,it can't parse text.
        //we could use co-body test it.
        // If koa has already parsed a body as a keyed object, use it.
        if (typeof ctx.req.body === 'object' && !(ctx.req.body instanceof Buffer)) {
            return resolve(ctx.req.body);
        }

        if (typeof ctx.request.body === 'object' && !(ctx.request.body instanceof Buffer)) {
            return resolve(ctx.request.body);
        }

        // Skip requests without content types.
        if (ctx.req.headers['content-type'] === undefined) {
            return resolve({});
        }

        const typeInfo = contentType.parse(ctx.req);

        // If koa has already parsed a body as a string, and the content-type
        // was application/graphql, parse the string body.
        if (typeof ctx.request.body === 'string' &&
            typeInfo.type === 'application/graphql') {
            return resolve(graphqlParser(ctx.request.body));
        }

        // Already parsed body we didn't recognise? Parse nothing.
        if (ctx.request.body) {
            return resolve({});
        }

        // Use the correct body parser based on Content-Type header.
        switch (typeInfo.type) {
            case 'application/graphql':
                return read(ctx.req, typeInfo, graphqlParser, resolve, reject);
            case 'application/json':
                return read(ctx.req, typeInfo, jsonEncodedParser, resolve, reject);
            case 'application/x-www-form-urlencoded':
                return read(ctx.req, typeInfo, urlEncodedParser, resolve, reject);
        }

        // If no Content-Type header matches, parse nothing.
        return resolve({});
    });
}

function jsonEncodedParser(body) {
    if (jsonObjRegex.test(body)) {
        /* eslint-disable no-empty */
        try {
            return JSON.parse(body);
        } catch (error) {
            // Do nothing
        }
        /* eslint-enable no-empty */
    }
    throw httpError(400, 'POST body sent invalid JSON.');
}

function urlEncodedParser(body) {
    return querystring.parse(body);
}

function graphqlParser(body) {
    return { query: body };
}

/**
 * RegExp to match an Object-opening brace "{" as the first non-space
 * in a string. Allowed whitespace is defined in RFC 7159:
 *
 *     x20  Space
 *     x09  Horizontal tab
 *     x0A  Line feed or New line
 *     x0D  Carriage return
 */
const jsonObjRegex = /^[\x20\x09\x0a\x0d]*\{/;

// Read and parse a request body.
function read(req, typeInfo, parseFn, resolve, reject) {
    const charset = (typeInfo.parameters.charset || 'utf-8').toLowerCase();

    // Assert charset encoding per JSON RFC 7159 sec 8.1
    if (charset.slice(0, 4) !== 'utf-') {
        throw httpError(415, `Unsupported charset "${charset.toUpperCase()}".`);
    }

    // Get content-encoding (e.g. gzip)
    const encoding =
        (req.headers['content-encoding'] || 'identity').toLowerCase();
    const length = encoding === 'identity' ? req.headers['content-length'] : null;
    const limit = 100 * 1024; // 100kb
    const stream = decompressed(req, encoding);

    // Read body from stream.
    getBody(stream, { encoding: charset, length, limit }, function (err, body) {
        if (err) {
            return reject(
                err.type === 'encoding.unsupported' ?
                    httpError(415, `Unsupported charset "${charset.toUpperCase()}".`) :
                    httpError(400, `Invalid body: ${err.message}.`)
            );
        }

        try {
            // Decode and parse body.
            return resolve(parseFn(body));
        } catch (error) {
            return reject(error);
        }
    });
}

// Return a decompressed stream, given an encoding.
function decompressed(req, encoding) {
    switch (encoding) {
        case 'identity': return req;
        case 'deflate': return req.pipe(zlib.createInflate());
        case 'gzip': return req.pipe(zlib.createGunzip());
    }
    throw httpError(415, `Unsupported content-encoding "${encoding}".`);
}

src/index.js
import {
    Source,
    parse,
    validate,
    execute,
    formatError,
    getOperationAST,
    specifiedRules
} from 'graphql';
import httpError from 'http-errors';
import url from 'url';
import { parseBody } from './parseBody';
import { renderGraphiQL } from './renderGraphiQL';

export default function graphqlHTTP(options) {
    if (!options) {
        throw new Error('GraphQL middleware requires options.');
    };
    return async (ctx, next) => {

        // Higher scoped variables are referred to at various stages in the
        // asyncronous state machine below.
        let schema;
        let context;
        let rootValue;
        let pretty;
        let graphiql;
        let formatErrorFn;
        let showGraphiQL;
        let query;
        let variables;
        let operationName;
        let validationRules;

        let optionsData;
        let result; //the content of response.body
        try {

            // Resolve the Options to get OptionsData.
            optionsData = await Promise.resolve(
                typeof options === 'function' ? options(ctx) : options
            );

            // Assert that optionsData is in fact an Object.
            if (!optionsData || typeof optionsData !== 'object') {
                throw new Error(
                    'GraphQL middleware option function must return an options object ' +
                    'or a promise which will be resolved to an options object.'
                )
            };

            // Assert that schema is required.
            if (!optionsData.schema) {
                throw new Error(
                    'GraphQL middleware options must contain a schema.'
                );
            };

            // Collect information from the options data object.
            schema = optionsData.schema;
            context = optionsData.context || ctx;
            rootValue = optionsData.rootValue;
            pretty = optionsData.pretty;
            graphiql = optionsData.graphiql;
            formatErrorFn = optionsData.formatError;

            validationRules = specifiedRules;
            if (optionsData.validationRules) {
                validationRules = validationRules.concat(optionsData.validationRules);
            }


            // GraphQL HTTP only supports GET and POST methods.
            if (ctx.request.method !== 'GET' && ctx.request.method !== 'POST') {
                ctx.response.set('Allow', 'GET, POST');
                throw httpError(405, 'GraphQL only supports GET and POST requests.');
            }

            //now parse body;
            let bodyData = await parseBody(ctx);


            result = await new Promise(resolve => {

                const urlData = ctx.request.url && url.parse(ctx.request.url, true).query || {};
                showGraphiQL = graphiql && canDisplayGraphiQL(ctx.request, urlData, bodyData);

                // Get GraphQL params from the request and POST body data.
                const params = getGraphQLParams(urlData, bodyData);
                query = params.query;
                variables = params.variables;
                operationName = params.operationName;

                // If there is no query, but GraphiQL will be displayed, do not produce
                // a result, otherwise return a 400: Bad Request.
                if (!query) {
                    if (showGraphiQL) {
                        resolve(null);
                    }
                    throw httpError(400, 'Must provide query string.');
                }

                // GraphQL source.
                const source = new Source(query, 'GraphQL request');

                // Parse source to AST, reporting any syntax error.
                let documentAST;
                try {
                    documentAST = parse(source);
                } catch (syntaxError) {
                    // Return 400: Bad Request if any syntax errors errors exist.
                    ctx.response.status = 400;
                    resolve({ errors: [syntaxError] });
                }

                // Validate AST, reporting any errors.
                const validationErrors = validate(schema, documentAST, validationRules);
                if (validationErrors.length > 0) {
                    // Return 400: Bad Request if any validation errors exist.
                    ctx.response.status = 400;
                    resolve({ errors: validationErrors });
                }

                // Only query operations are allowed on GET requests.
                if (ctx.request.method === 'GET') {
                    // Determine if this GET request will perform a non-query.
                    const operationAST = getOperationAST(documentAST, operationName);
                    if (operationAST && operationAST.operation !== 'query') {
                        // If GraphiQL can be shown, do not perform this query, but
                        // provide it to GraphiQL so that the requester may perform it
                        // themselves if desired.
                        if (showGraphiQL) {
                            resolve(null);
                        }

                        // Otherwise, report a 405: Method Not Allowed error.
                        ctx.response.set('Allow', 'POST');
                        throw httpError(
                            405,
                            `Can only perform a ${operationAST.operation} operation ` +
                            'from a POST request.'
                        );
                    }
                }
                // Perform the execution, reporting any errors creating the context.
                try {
                    resolve(execute(
                        schema,
                        documentAST,
                        rootValue,
                        context,
                        variables,
                        operationName
                    ));
                } catch (contextError) {
                    // Return 400: Bad Request if any execution context errors exist.
                    ctx.response.status = 400;
                    resolve({ errors: [contextError] });
                };
            }); //end promise
        } catch (error) {
            // If an error was caught, report the httpError status, or 500.
            ctx.response.status = error.status || 500;
            result = { errors: [error] };
        }

        // Format any encountered errors.
        if (result && result.errors) {
            result.errors = result.errors.map(formatErrorFn || formatError);
        }
        // If allowed to show GraphiQL, present it instead of JSON.
        if (showGraphiQL) {
            const data = renderGraphiQL({
                query, variables,
                operationName, result
            });
            ctx.response.type = 'text/html; charset=utf-8';
            ctx.response.body = data;
        } else {
            // Otherwise, present JSON directly.
            const data = JSON.stringify(result, null, pretty ? 2 : 0);
            ctx.response.type = 'application/json; charset=utf-8';
            ctx.response.body = data;
        };
        next(); //add next here?

    }; //end middleware
}; //end graphqlHTTP


/**
 * Helper function to get the GraphQL params from the request.
 */
function getGraphQLParams(urlData, bodyData) {
    // GraphQL Query string.
    const query = urlData.query || bodyData.query;
    // Parse the variables if needed.
    let variables = urlData.variables || bodyData.variables;
    if (variables && typeof variables === 'string') {
        try {
            variables = JSON.parse(variables);
        } catch (error) {
            throw httpError(400, 'Variables are invalid JSON.');
        }
    }

    // Name of GraphQL operation to execute.
    const operationName = urlData.operationName || bodyData.operationName;
    return { query, variables, operationName };
}

/**
 * Helper function to determine if GraphiQL can be displayed.
 */
function canDisplayGraphiQL(
    request,
    urlData,
    bodyData
) {
    // If `raw` exists, GraphiQL mode is not enabled.
    const raw = urlData.raw !== undefined || bodyData.raw !== undefined;
    // Allowed to show GraphiQL if not requested as raw and this request
    // prefers HTML over JSON.
    //modify:using request.accepts instead od accepts.
    return !raw && request.accepts(['json', 'html']) === 'html'; //fix:test fail.
}
